crates/admin-rest-model/src/version.rs

```diff
@@ -23,13 +23,46 @@ use std::ops::RangeInclusive;
 pub enum AdminApiVersion {
     Unknown = 0,
     V1 = 1,
+    V2 = 2,
+}
+
+impl From<AdminApiVersion> for (http::HeaderName, http::HeaderValue) {
+    fn from(value: AdminApiVersion) -> Self {
+        (
+            AdminApiVersion::HEADER_NAME,
+            http::HeaderValue::from(value.as_repr()),
+        )
+    }
 }
 
 impl AdminApiVersion {
+    const HEADER_NAME: http::HeaderName =
+        http::HeaderName::from_static("x-restate-admin-api-version");
+
     pub fn as_repr(&self) -> u16 {
         *self as u16
     }
 
+    pub fn from_headers(headers: &http::HeaderMap) -> Self {
+        let is_cli = matches!(headers.get(http::header::USER_AGENT), Some(value) if value.as_bytes().starts_with(b"restate-cli"));
+
+        match headers.get(Self::HEADER_NAME) {
+            Some(value) => match value.to_str() {
+                Ok(value) => match value.parse::<u16>() {
+                    Ok(value) => match Self::try_from(value) {
+                        Ok(value) => value,
+                        Err(_) => Self::Unknown,
+                    },
+                    Err(_) => Self::Unknown,
+                },
+                Err(_) => Self::Unknown,
+            },
+            // CLI didn't used to send the version header, but if we know its the CLI, then we can treat that as V1
+            None if is_cli => Self::V1,
+            None => Self::Unknown,
+        }
```

Comment:
Personally, I am not a huge fan of this kind of special casing because it can be confusing. 

Is it standard practice to consider a request w/o a version header as the latest? While this is very convenient, it seems also like the best way to run into breaking changes because initially users don't have to think about versions until they upgrade.

Would it be an option to always consider a missing version header as v1?

Reply:
i think it is not standard practice. i think if its a header, we need to make sure that the naive curl case without the header does something sensible. if its a url path, we could perhaps deprecate the unversioned one, but given that we are versioning the whole api, it could mean fairly regularly breaking old scripts/clients which use a versioned url, when we remove support for an old api version. i don't really have a great answer here!

Reply:
Yeah, the versioning approach via headers makes it a bit more hidden and I agree that the basic curl case should still work.

Regarding deprecating a versioned url, I wouldn't worry too much about it. This is something that we can announce and then slowly phase out to give users time to adjust.